### PR TITLE
Inter-module Communication

### PR DESCRIPTION
**Pull request by tafseer-khan**
I2mv or Inter-Module-Memory-Viewer, allows creation of a closer on one module, to be accessed by another module as a read-seek-closer. 

Current testing is between 2 golang modules, but the intention is to use this mechanism between a golanag module and another module of different language. 

sdk mock tests and examples: 
https://asciinema.org/a/49xmjdwnuypYejJQXv2vipEUm

vm-test-examples:
https://asciinema.org/a/BgAX0sIKeRNnm9WQfAY5Ow0xO

### CODE REVIEW

File: i2mv/new.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	}
+
+	var closable uint32
+	if readCloser {
+		closable = 1
+	}
+
+	err0 := symbols.MemoryViewNew(&data[0], uint32(len(data)), closable, &id)
+	if err0 != 0 {
+		return 0, nil, fmt.Errorf("creating new memory view failed with: %s", err)
```
Comment (skelouse) [inline i2mv/new.go:22]: failed with err0

File: i2mv/methods.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	return m.offset
+}
+
+func (m *memoryView) Size() uint32 {
+	return m.size
+}
+
+func (m *memoryView) Read(p []byte) (int, error) {
+	var n uint32
+	if p == nil || len(p) == 0 {
```
Comment (skelouse) [inline i2mv/methods.go:32]: []byte{nil}  will have len == 0

File: i2mv/i2mv_test.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	mv, err := Open(testId)
+	if err != nil {
+		t.Errorf("opening memory view failed with: %s", err)
+		return
+	}
+
+	symbols.MockClose(true)
+
+	if err := mv.Close(); err != nil {
+		t.Errorf("closing memory viewfailed with: %s", err)
```
Comment (skelouse) [inline i2mv/i2mv_test.go:165]: ` `

File: i2mv/methods.go (hunk 2 above)
Comment (tafseer-khan) [reply i2mv/methods.go:32]: There was a weird use case aron showed me a while ago where bytes were not nil, but with length 0

File: i2mv/i2mv_test.go (hunk 3 above)
Comment (tafseer-khan) [reply i2mv/i2mv_test.go:165]: what

File: errno/generate/main.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	"path"
+	"strings"
+)
+
+func main() {
+	wd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+	readFile, err := os.Open(path.Join(wd, "types.go"))
```
Comment (taubyte-test) [inline errno/generate/main.go:16]: new line before

File: i2mv/methods.go (hunk 2 above)
Comment (taubyte-test) [reply i2mv/methods.go:32]: []byte == nil, []byte{} == 0

Comment (skelouse) [reply i2mv/methods.go:32]: Then just check len == 0

File: i2mv/i2mv_test.go (hunk 3 above)
Comment (skelouse) [reply i2mv/i2mv_test.go:165]: View failed

File: errno/generate/main.go (hunk 4 above)
Comment (arontaubyte) [inline errno/generate/main.go:16]: new line before

File: i2mv/methods.go (hunk 2 above)
Comment (arontaubyte) [reply i2mv/methods.go:32]: []byte == nil []byte{} == 0

File: i2mv/methods.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	case io.SeekCurrent:
+		_whence = m.offset
+	case io.SeekEnd:
+		_whence = int64(m.size)
+	default:
+		return 0, fmt.Errorf("invalid whence value `%d`", whence)
+	}
+
+	if _offSet := _whence + offSet; _offSet < 0 {
+		return 0, fmt.Errorf("cannot seek before start")
```
Comment (arontaubyte) [inline i2mv/methods.go:59]: errors.New

File: i2mv/new.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"fmt"
+	"io"
+
+	symbols "github.com/taubyte/go-sdk-symbols/i2mv"
+)
+
+func New(data []byte, readCloser bool) (id uint32, closer io.Closer, err error) {
+	if data == nil || len(data) == 0 {
+		return 0, nil, fmt.Errorf("cannot create memory view from nil data")
```
Comment (arontaubyte) [inline i2mv/new.go:12]: errors.New

**Review by skelouse — APPROVED** [review]

**Review by arontaubyte — APPROVED** [review]

**Pull request merged**